source/common/protobuf/utility.h

```diff
@@ -177,6 +178,26 @@ class MessageUtil {
   static void loadFromYaml(const std::string& yaml, Protobuf::Message& message);
   static void loadFromFile(const std::string& path, Protobuf::Message& message);
 
+  template <class MessageType>
+  static void checkForDeprecation(const MessageType& message, bool fatal) {
+    const auto* desc = message.GetDescriptor();
+    const auto* refl = message.GetReflection();
+    for (int i = 0; i < desc->field_count(); ++i) {
```

Comment:
This only checks top level field and not about inner fields such as [here](https://github.com/envoyproxy/envoy/pull/5760/files#diff-e85fbd6cdad5d62882b56ccdac06ff45R9), Validate does recursive validation internally but MessageUtil::validate isn't called at every level.

Reply:
Ah my bad, I didn't realize Validate was doing the recursion - will fix and regression test